certora/specs/Delegation.spec

```diff
@@ -0,0 +1,77 @@
+import "ERC20.spec";
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sload uint256 votingPower (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] {
+    require sumOfVotingPower >= to_mathint(votingPower);
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] uint256 newValue (uint256 oldValue) {
+    sumOfVotingPower = sumOfVotingPower - oldValue + newValue;
+}
+
+// Check that zero address has no voting power assuming that zero address can't make transactions.
+invariant zeroAddressNoVotingPower()
+    delegatee(0x0) == 0x0 && delegatedVotingPower(0x0) == 0
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    { preserved with (env e) { require e.msg.sender != 0; } }
+
+// Check that the voting power plus the virtual voting power of address zero is equal to the total supply of tokens.
+invariant totalSupplyIsSumOfVirtualVotingPower()
+    to_mathint(totalSupply()) == sumOfVotingPower + currentContract._zeroVirtualVotingPower
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    {
+      preserved {
+          requireInvariant totalSupplyIsSumOfBalances();
+          requireInvariant zeroAddressNoVotingPower();
+      }
+    }
+
+// Check that the total supply of tokens is greater than or equal to the sum of voting power.
+invariant totalSupplyGTEqSumOfVotingPower()
```

Comment:
Indeed, you're right. However, if I remember correctly our previous discussion, the whole point of this invariant is to be able to write `requireInvariant` such that invariant required is exactly the invariant proven, by contrast with how it's implemented. Namely, this invariant is required in the spec `delegatingUpdatesVotingPower`, AFAIK it's not possible to require rules.

Reply:
I wouldn't say it's the whole point of this invariant. It's more: if it is an invariant, then we have to make the name be representative of the formula proven, because otherwise it could be confusing when using `requireInvariant`.
But you are right that it's not possible to require a rule, so it's a bit annoying to change it here. So we could use one of the following solutions, I think all of them could make sense and I let you choose between them:
- use a function to describe the assert in the rule, and require that function later instead of `requireInvariant delegatingUpdatesVotingPower`
- require the invariant `totalSupplyIsSumOfVirtualVotingPower` in place where `delegatingUpdatesVotingPower` is required, the prover should be able to deduce the rest
- don't change anything

Reply:
I like the idea with the function. Implemented https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/873cfdec214ae3025c8163335f1afd632c1d98a7.